code for preparing cells dict got moved from global level into the function loadCells

mineserver228 opened the pull request.

mineserver228 committed:
code that was preparing cells dict got moved from global level into the function loadCells

mineserver228 committed:
fixed small bugs in loadEntropy from Cell class

kolayne commented on wfcGen.py:
```diff
@@ -2,83 +2,79 @@
 from PIL import Image
 
+def loadCells(imageLink, tileSize=3, cropShift=False):
+    cropShift = cropShift
```
??

kolayne commented on wfcGen.py:
```diff
@@ -2,83 +2,79 @@
+def loadCells(imageLink, tileSize=3, cropShift=False):
+    cropShift = cropShift
+    cells = {}
+    cropped = []
+    for ins in range(4):
+        print(ins/4)
+        field = Image.open(imageLink)
+        cropped.append([])
```
It appears to me that this function always returns `[[], [], [], []]`. Why?

mineserver228 replied:
there is if else after this
on 30 and 38 it sets cropped to something

kolayne replied:
Sorry, didn't read this right, for some reason, it appeared to me there is a return statement after the loop. 

Still, what is the purpose of this loop? You can avoid it completely, can't you?

kolayne commented on wfcGen.py:
```diff
@@ -2,83 +2,79 @@
+            copy.paste(field, (SX, 0))
+            copy.paste(field, (0, SY))
+            copy.paste(field, (SX, SY))
+            field = copy.copy()
```
Why do you create something called `copy` and then copy it to `field`, if you don't ever use `copy` again...

kolayne commented on wfcGen.py:
```diff
@@ -2,83 +2,79 @@
-                    flag = key
+        for y in range(lcy):
+            for x in range(lcx):
+                flag = -1
```
Bad variable name

kolayne commented on wfcGen.py:
```diff
@@ -2,83 +2,79 @@
+                        Gkey = key
+                for key in cells.keys():
+                    for key2 in cells[key]["nbg"].keys():
+                        if key2 != "img":
```
1. It's probably not the best option to name these iterators `key` and `key2`, as it doesn't add much information. Can you instead name them to reflect something about what value they hold?
2. Maybe this piece of code (and a littlbe bit following it) would look cleaner if used `dict.items()` rather than `dict.keys()` in this case:
```python
d = {...: ..., ...}

# Option 1
for k in d.keys():
    do_something_with(d[k])

# Option 2
for k, v in d.items():
    do_something_with(v)
```

(please, remember to name both `k` and `v` meaningfully in your actual code)

mineserver228 replied:
i won't use this everywhere in the code, because sometimes i need to do something with "d"

kolayne commented on wfcGen.py:
```diff
@@ -93,7 +89,7 @@ def __init__(self, gp):
         self.checkCollapsedNeighbours = True
 
-    def collapse(self):
+    def collapse(self, cells):
```
Notice how you had to add the `cells` parameter to almost every method. This is a bad practice and it might suggest that your architecture is not well-structured enough.

In general, in such a case I would consider the following points:
- Does it make sense if for this class I first call one method with one value of `cells`, and then call another method with another value of `cells`. If it does, then they should probably remain method parameters. If it does not, something is wrong.
- Does it make sense to make `cells` a field of the class rather than accepting it as a method argument every time?

In this particular case there is another thing that draws my attention: from a single `Cell` you are trying to access the array of (all?) cells. This is suspicious. Do you really need that? Maybe you only need some particular information from that array?

mineserver228 replied:
I think i will make Grid class to take cells as a parameter when created
class Cell requiers whole cells dictionary to collapse, but I will fix getEntropy function, so it will not use cells dict

kolayne replied:
Alright. Didn't get all of what you said but let's try to do that

kolayne commented on wfcGen.py:
```diff
@@ -211,35 +207,38 @@ def collapse(self):
-                    self.GIMAGE.paste(cells[self.grid[y][x].id]["img"], (x*res, y*res, x*res + res, y*res + res))
-        self.GIMAGE.save("result.png")
+                    self.GIMAGE.paste(cells[self.grid[y][x].id]["img"], (x*tileSize, y*tileSize, x*tileSize + tileSize, y*tileSize + tileSize))
+        self.GIMAGE.save("tileSizeult.png")
```
Hardcoded filename in a class method

mineserver228 committed:
code look and logic changes

- changed some variables names
  key -> CellIdKey
  key2 -> neighborKey
  flag -> cellId

- changed class Grid logic
  - now it only takes cells dict when created
  - function getFinalImage now takes optional saveFileName parameter

- changed class Cell logic
 - function getEntropy now doesn't require cells dict

- deleted some unnecesary code

- made some small code look and logic improvement

kolayne commented on wfcGen.py:
```diff
@@ -2,83 +2,79 @@
 
+def loadCells(imageLink, tileSize=3, cropShift=False):
+    cells = {}
+    cropped = []
+    startImage = Image.open(imageLink)
+    for ins in range(4):
+        print(ins/4)
+        
+        cropped.append([])
```
As I can see, on every iteration you insert an empty array here, and then overwrite it later by writing something to `cropped[ins]`. Does it make sense to remove the `append` here and, instead, do `cropped.append` in the end, when you have data to append?

kolayne commented on wfcGen.py:
```diff
@@ -2,83 +2,79 @@
+                else:
+                    cells[len(cells.keys())] = {
+                        "nbg" : {
+                            0 : [cropped[ins][y][x-1-(tileSize-1)*int(cropShift)]],
+                            1 : [cropped[ins][y-1-(tileSize-1)*int(cropShift)][x]],
+                            2 : [cropped[ins][y][(x+1+(tileSize-1)*int(cropShift))%lcx]],
+                            3 : [cropped[ins][(y+1+(tileSize-1)*int(cropShift))%lcy][x]],
```
I suggest that rather than using indexing, you use names for the directions in which the cells are adjacent. I think a `NamedTuple` fits well here

mineserver228 replied:
wait, you can not actually change tuples, can you? because I change value of 0, 1, 2 and 3 than
so I think I will just change names from numbers to strings like "left", "top" ...

kolayne replied:
Have a look at [this SO question](https://stackoverflow.com/q/42618134/11248508). One of the answers gives an example of how one can work with `NamedTuple`s if they want to change their values, as well as suggests an alternative which allows for mutable values similar to `NamedTuple` in how you access them

kolayne commented on wfcGen.py:
```diff
@@ -134,24 +130,24 @@ def loadEntropy(self, grid):
         else:
             if self.checkCollapsedNeighbours:
-                return len(cells.keys())
+                return 9223372036854775800
```
??
Please, at least write a comment to explain
(maybe for the `-1` return value as well)

kolayne reviewed (commented):
Top, the code looks better now!

But there are still things to fix. Here they are

mineserver228 committed:
several bug fixes

mineserver228 committed:
small code structure changes

- dict cells structure has been chaged
  - in "nbg" key now "left", "top", ..., instead of
    0,1,2,3

- deleted some unnecesary code

kolayne commented on wfcGen.py:
```diff
@@ -112,14 +111,15 @@ def loadEntropy(self, grid):
                 if (self.gp[1]+xy)<len(grid):
                     if grid[self.gp[1]+xy][self.gp[0]].collapsed:
-                        self.entropy = grid[self.gp[1]+xy][self.gp[0]].nbg[2 - xy]
+                        self.entropy = grid[self.gp[1]+xy][self.gp[0]].nbg[list(grid[self.gp[1]+xy][self.gp[0]].nbg.keys())[2 - xy]]
```
This is completely unreadable

mineserver228 committed:
code logic changes

- function loadEntropy in class Cell survived throught big changes

- code became more readable

kolayne commented on wfcGen.py:
```diff
@@ -145,13 +143,13 @@ def getEntropy(self):
-            if self.checkCollapsedNeighbours:
-                return len(cells.keys())
+            if self.collapsedNeighbours:
+                return 9223372036854775800 # max int - 1, this number need to be bigger than len of cells
```
Firstly, that isn't technically guaranteed that a Python list length is less than int max (because Python has long arithmetics), although this situation appears extremely unlikely because I doubt anyone has this much RAM.

Still, how about
```suggestion
                return len(cells) + 1
```

kolayne commented on wfcGen.py:
```diff
@@ -2,83 +2,78 @@
-
-    print(cells)
+                if cellId != -1:
+                    cells[cellId]["nbg"]["left"].append(cropped[ins][y][x-1-(tileSize-1)*int(cropShift)])
+                    cells[cellId]["nbg"]["top"].append(cropped[ins][y-1-(tileSize-1)*int(cropShift)][x])
+                    cells[cellId]["nbg"]["right"].append(cropped[ins][y][(x+1+(tileSize-1)*int(cropShift))%lcx])
+                    cells[cellId]["nbg"]["bottom"].append(cropped[ins][(y+1+(tileSize-1)*int(cropShift))%lcy][x])
+                else:
+                    cells[len(cells.keys())] = {
+                        "nbg" : {
+                            "left" : [cropped[ins][y][x-1-(tileSize-1)*int(cropShift)]],
+                            "top" : [cropped[ins][y-1-(tileSize-1)*int(cropShift)][x]],
+                            "right" : [cropped[ins][y][(x+1+(tileSize-1)*int(cropShift))%lcx]],
+                            "bottom" : [cropped[ins][(y+1+(tileSize-1)*int(cropShift))%lcy][x]],
```
These expressions are still hard to read. You can definitely improve them by extracting some subexpressions (e.g., `(tileSize - 1) * int(cropShift)`? Anything else?) into separate variables with names. This will both give more insight into what's going on (because there is a named variable), and simplify the index expressions

kolayne reviewed (changes requested):
Makes overall sense to me, although there still are some issues

mineserver228 committed:
mall code changes

- Changed functions
  - loadCells - small changes to make it look better
  - getEntropy - maxInt-1 changed to len(cells), not to len(cells) + 1
    for technical reasons

mineserver228 committed:
deleted unncesary comment
